WIP: Replicate stack gan

Devin-Taylor opened the pull request:
DO NOT MERGE

Devin-Taylor commented on models/generators.py:
```diff
@@ -33,15 +35,18 @@ def __init__(self, img_size, num_latent_dims, kernel_size, num_filters,
     @tf.function
-    def call(self, noise):
-        x = self.dense1(noise)
+    def call(self, noise, embedding):
```
@alecokas can I assume embedding will be passed in here?

alecokas replied:
Yeah I think that makes sense

alecokas commented on models/conditional_gans.py:
```diff
@@ -65,13 +68,98 @@ class StackGAN1(ConditionalGAN):
+            img_size=img_size,
+            kernel_size=kernel_size,
+            num_filters=num_filters
+        )
```
Why are these before calling `super`?

Devin-Taylor replied:
they need to be passed into super because each stage has different generators and discriminators. Tbh I do not know if we really need the parent class here, but I just left it for now.

alecokas replied:
OK, sounds good :+1:

alecokas commented on utils/data_helpers.py:
```diff
@@ -137,11 +140,21 @@ def download_flowers():
+                              (image_dims[1]-new_size[1])//2))
+        img_buffer = io.BytesIO()
+        new_img.save(img_buffer, format='PNG')
+        image_string = img_buffer.getvalue()
```
I assume this is the code that maintains the aspect ratio?

Devin-Taylor replied:
yup, pretty much:
1) determine aspect ratio
2) determine reduced image size required to maintain ar
3) create blank square image
4) paste resized image in middle of blank image

alecokas replied:
:+1:

alecokas commented on utils/utils.py:
```diff
@@ -53,7 +53,7 @@ def sample_normal(mean, log_var):
     """
     std = tf.math.exp(0.5 * log_var)
-    epsilon = tf.random.normal(std)
+    epsilon = tf.random.normal(tf.shape(std)) # TODO is shape the correct thing to use here?
```
What do you mean?

Devin-Taylor replied:
I think it is ok, but I wanted to double check with the paper when I had a chance

alecokas replied:
Checked: this is correct

alecokas commented on utils/utils.py:
```diff
@@ -82,7 +82,8 @@ def save_options(options, save_dir):
 
-    mkdir(save_dir)
+    if not os.path.exists(save_dir):
+        mkdir(save_dir)
```
`mkdir` has a check built into it, so you don't need to check before calling the function

Devin-Taylor replied:
ah true, will edit it

Devin-Taylor commented:
@alecokas please review and merge when happy. Stage 1 implementation complete and running end-to-end. Please *do not* delete the branch.

Changes:
- resize images, maintaining aspect ratio, pad border to get square
- implemented stage 1 conditional augmentation
- implemented stage 1 generator
- implementation stage 1 discriminator
- implement generator and discriminator loss functions
- configured training loop
- integrated batched training

Notes:
- Not training perfectly (discriminator winning too easily) but will resolve that in a different PR.

alecokas commented on models/generators.py:
```diff
@@ -50,4 +54,115 @@ def loss(self, predictions_on_fake):
+        return self.xent_loss_fn(predictions_on_fake, tf.ones_like(predictions_on_fake)) + kl_coeff * self.kl_loss(mean, log_sigma)
+
+    def kl_loss(self, mean, log_sigma):
+        loss = -log_sigma + .5 * (-1 + tf.exp(2. * log_sigma) + tf.square(mean))
```
Assuming this is correct for now lol

alecokas commented on trainers/text_to_image_trainer.py:
```diff
@@ -1,7 +1,10 @@
 import io
+
 import numpy as np
+import tensorflow as tf
 from PIL import Image
 from tqdm import trange
+
 from trainers.base_trainer import Trainer
 
```
Why the spaces?

Devin-Taylor replied:
right click + sort imports, do not as me ¯\_(ツ)_/¯

alecokas replied:
:joy:

Devin-Taylor commented on models/discriminators.py:
```diff
@@ -48,4 +52,111 @@ def loss(self, predictions_on_real, predictions_on_fake):
+        self.conv1 = Conv2D(filters=64, kernel_size=(4, 4), strides=(2, 2))
+        self.leaky_relu_1 = LeakyReLU()
+
+        self.conv2 = Conv2D(filters=64*2, kernel_size=(4, 4), strides=(2, 2))
```
haha na it shouts at me for that, you using vscode default python extension?

alecokas replied:
I usually use flake8 - I'm fine with this as is. Just curious because flake8 complains I think

alecokas commented on models/conditional_gans.py:
```diff
@@ -65,13 +68,98 @@ class StackGAN1(ConditionalGAN):
+                learnt log variance of embedding distribution. Shape (batch_size, reshape_dims/2)
+
+        """
+        x = self.flatten(embedding)
```
So we take the mean and var over all elements? Across all samples, all features, and all embedding dimensions?

Devin-Taylor replied:
yeah, technically. Might be worth averaging over the sample dimension first then flattening?

alecokas replied:
Yeah we should double check this

Devin-Taylor replied:
hard to tell, discussions on the multiple samples are vague. I think there was an older paper that mentions it, will dig it up and see what they do

alecokas commented on models/conditional_gans.py:
```diff
@@ -65,13 +68,98 @@ class StackGAN1(ConditionalGAN):
+        sigma = self.dense_sigma(x)
+        return mean, sigma
+
+    def conditional_augmentation(self, embedding):
+        """ Perform conditional augmentation by sampling embedding.
+            Arguments
+            embedding : Tensor
+                text embedding. Shape (batch_size, feature_size, embedding_size)
+
+            Returns
+            Tensor
+                sampled embedding. Shape (batch_size, reshape_dims/2)
+        """
+        mean, sigma = self.generate_conditionals(embedding)
+        smoothed_embedding = sample_normal(mean, sigma)
+        return smoothed_embedding, mean, sigma
```
So do I understand correctly that conditional augmentation is simply sampling a normal with the mean and var of all the elements across all dims in the batch? I'm asking because I can't really see a random variable that is being conditioned on. I will have to re-read the papers, but I thought worth checking because it doesn't seem intuitive to me right now

Devin-Taylor replied:
![Screenshot from 2020-04-06 23-21-53](https://user-images.githubusercontent.com/13699844/78606364-7ea11900-785d-11ea-8db9-9e86b3dfa054.png)

alecokas commented on trainers/text_to_image_trainer.py:
```diff
@@ -30,16 +33,54 @@ def train_epoch(self, train_loader, epoch_num):
+                noise_z = tf.random.normal([batch_size, 100])
+
+                with tf.GradientTape() as generator_tape, tf.GradientTape() as discriminator_tape:
+                    smoothed_embedding, mean, log_sigma = self.model.conditional_augmentation(text_tensor)
+                    embedding_z = tf.concat([smoothed_embedding, noise_z], 1)
```
Should we be doing this outside the gradient tape maybe?

alecokas replied:
Oh wait never mind, we have some linear layers in there that need to be backpropped.

Devin-Taylor replied:
conditional augmentation has the linear layers, so are trainable. Which means they need to be on the tape, no?

alecokas commented on models/conditional_gans.py:
```diff
@@ -65,13 +68,98 @@ class StackGAN1(ConditionalGAN):
+    @tf.function
+    def call(self, embedding, z):
+        """
+        """
```
Maybe just add `None` or raise `NotImplementedError` - I think this is a syntax error atm.

Devin-Taylor replied:
cool, fixed
